### PR TITLE
3477 deploy ttapi govuk paas GitHub actions

### PR DESCRIPTION
**Pull request by karthikdfe**
### Checklist

- [ ] Make sure all information from the Trello card is in here
- [ ] Attach to Trello card
- [ ] Rebased master
- [ ] Cleaned commit history
- [ ] Tested by running locally

### CODE REVIEW

File: terraform/paas/resources.tf [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+  domain   = data.cloudfoundry_domain.local.id
+  space    = data.cloudfoundry_space.space.id
+  hostname = var.app.hostname
+}
+
+resource "cloudfoundry_service_instance" "postgres" {
+  name         = var.paas_postgres_name
+  space        = data.cloudfoundry_space.space.id
+  service_plan = data.cloudfoundry_service.postgres.service_plans["small-11"]
+  json_params  = "{\"enable_extensions\": [\"postgis\"] }"
```
Comment (asmega) [inline terraform/paas/resources.tf:39]: Do we need postgis? I don't think we use it. According to the schema https://github.com/DFE-Digital/teacher-training-api/blob/master/db/schema.rb#L16 we need
```
 16   enable_extension "pg_buffercache"
 17   enable_extension "pg_stat_statements"
 18   enable_extension "plpgsql"
```
although I'd have to check if they are addons or part of the postgres packages

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    steps:
+    - name: Checkout the repository
+      uses: actions/checkout@v2
+    - name: Pull any existing images for this branch to cache
+      run: |
+        docker pull $DOCKER_PATH:$GIT_BRANCH || true
+        docker pull $DOCKER_PATH-middleman:master || true
+    - name: Build docker image
+      run: |
+        sudo $DOCKER_COMPOSE build middleman
```
Comment (asmega) [inline .github/workflows/build.yml:36]: why does docker-compose need sudo?

**Comment by misaka** [comment]
Feels like this PR should still be in draft.

**Comment by misaka** [comment]
Oh look, there's a way to do that.

File: terraform/paas/terraform_qa.tfvars [hunk 3]
```diff
@@ -1,6 +1,6 @@
 app = {
   name               = "qa-teacher-training-api"
-  docker_image       = "dfedigital/teacher-training-api:testr"
+  docker_image       = "karthikdfe/teacher-training-api-paas"
```
Comment (asmega) [inline terraform/paas/terraform_qa.tfvars:3]: did we lose access to the dfe repo?

Comment (saliceti) [reply terraform/paas/terraform_qa.tfvars]: We have now created a specific user `finddockerhub` to push to the right repo.

File: .github/workflows/build.yml (hunk 2 above)
Comment (saliceti) [reply .github/workflows/build.yml]: Removed, thanks

File: terraform/paas/resources.tf (hunk 1 above)
Comment (saliceti) [reply terraform/paas/resources.tf]: Removed, thanks

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+        TF_VAR_SENTRY_DSN: ${{ secrets.TF_VAR_SENTRY_DSN }}
+        # Let's confirm if these variables are required
+        #TF_VAR_SETTINGS__GOOGLE__GCP_API_KEY: ${{ secrets.TF_VAR_SETTINGS__GOOGLE__GCP_API_KEY }}
+        #TF_VAR_SETTINGS__GOOGLE__MAPS_API_KEY: ${{ secrets.TF_VAR_SETTINGS__GOOGLE__MAPS_API_KEY }}
+        #TF_VAR_AUTHENTICATION_TOKEN: ${{ secrets.TF_VAR_AUTHENTICATION_TOKEN }}
+        #TF_VAR_SETTINGS__AUTHENTICATION__SECRET: ${{ secrets.SETTINGS__AUTHENTICATION__SECRET }}
+        #TF_VAR_SETTINGS__GOVUK_NOTIFY__API_KEY: ${{ secrets.SETTINGS__GOVUK_NOTIFY__API_KEY }}
+        #TF_VAR_SETTINGS__GOVUK_NOTIFY__COURSE_PUBLISH_EMAIL_TEMPLATE_ID: ${{ secrets.SETTINGS__GOVUK_NOTIFY__COURSE_PUBLISH_EMAIL_TEMPLATE_ID }}
+        #TF_VAR_SETTINGS__SETTINGS__GOVUK_NOTIFY__COURSE_UPDATE_EMAIL_TEMPLATE_ID: ${{ secrets.SETTINGS__GOVUK_NOTIFY__COURSE_UPDATE_EMAIL_TEMPLATE_ID }}
+        #TF_VAR_SETTINGS__GOVUK_NOTIFY__WELCOME_EMAIL_TEMPLATE_ID: ${{ secrets.SETTINGS__GOVUK_NOTIFY__WELCOME_EMAIL_TEMPLATE_ID }}
```
Comment (misaka) [inline .github/workflows/build.yml:140]: GOVUK Notify variables that are still required:

```ruby
Settings.govuk_notify.api_key
Settings.govuk_notify.course_publish_email_template_id
Settings.govuk_notify.course_sites_update_email_template_id
Settings.govuk_notify.course_subjects_updated_email_template_id
Settings.govuk_notify.course_update_email_template_id
Settings.govuk_notify.course_vacancies_updated_email_template_id
Settings.govuk_notify.course_withdraw_email_template_id
Settings.govuk_notify.magic_link_email_template_id
Settings.govuk_notify.welcome_email_template_id
```

We're working towards pulling these out of the code, but that's a separate exercise. I pulled this list from the code base in master:

```sh
ag Settings.govuk_notify app config | ruby -pe '$_.sub!(/.*(Settings.govuk_notify[.a-z0-9_-]+).*/, "\\1")' | sort -u
```

Comment (saliceti) [reply .github/workflows/build.yml:140]: OK let's add them. I assume only `api_key` is a secret?

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+      run: |
+        docker push $DOCKER_IMAGE:$SHA_TAG
+        docker push $DOCKER_IMAGE:$BRANCH_TAG
+        [[ "$BRANCH_TAG" == "master" ]] && docker push $DOCKER_IMAGE-middleman:$BRANCH_TAG
+        docker push $DOCKER_IMAGE_MAILER:$SHA_TAG
+        docker push $DOCKER_IMAGE_MAILER:$BRANCH_TAG
+        docker push $DOCKER_IMAGE_GEOCODE:$SHA_TAG
+        docker push $DOCKER_IMAGE_GEOCODE:$BRANCH_TAG
+        docker push $DOCKER_IMAGE_JOBS:$SHA_TAG
+        docker push $DOCKER_IMAGE_JOBS:$BRANCH_TAG
```
Comment (misaka) [inline .github/workflows/build.yml:91]: As discussed over Zoom, we'll remove these pushes and only push up the main image (and middleman), and then configure the `bg-jobs` background task to run off of the main image and specifying the command to run (`sidekiq`) on all the queues.

File: .github/workflows/build.yml (hunk 4 above)
Comment (asmega) [reply .github/workflows/build.yml:140]: @saliceti Yes `api_key` is a secret and should not be available. The others are merely identifiers and can be public knowledge. In the future we plan to move these identifiers out of the environment variable and just hard code them in the app.

**Comment by asmega** [comment]
[[L]Deploy ttapi app onto GOV.UK PAAS using GitHub Actions](https://trello.com/c/SyFcq8RJ/3477-ldeploy-ttapi-app-onto-govuk-paas-using-github-actions)

File: .github/workflows/build.yml (hunk 4 above)
Comment (felixtheflex) [reply .github/workflows/build.yml:140]: Having looked on the docker's environment on Azure and the deploy.json file. The following two keys cant be found:-

* Settings.govuk_notify.course_sites_update_email_template_id
* Settings.govuk_notify.course_subjects_updated_email_template_id

In light of this, are these two keys being used at all?

Comment (asmega) [reply .github/workflows/build.yml:140]: It looks like they have been hard coded here https://github.com/DFE-Digital/teacher-training-api/blob/master/config/settings.yml#L9 which is what we want to move to in the long term. So if they are deleted in azure this should not be a problem

**Comment by felixtheflex** [comment]
This is now ready to be merge to master... could this be reviewed please....?

File: .github/workflows/build.yml (hunk 5 above)
Comment (felixtheflex) [reply .github/workflows/build.yml:91]: This is now ready to be merge to master... could this be reviewed please....?

File: .github/workflows/build.yml [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+    - name: "docker pull ${{env.DOCKER_IMAGE}}-middleman:master"
+      run: docker pull $DOCKER_IMAGE-middleman:master || true
+
+    - name: "docker-compose build middleman"
+      run: $DOCKER_COMPOSE build middleman
+      env:
+        GIT_BRANCH: ${{env.BRANCH_TAG}}
+
+    - name: "docker-compose build"
+      run: $DOCKER_COMPOSE build
```
Comment (misaka) [inline .github/workflows/build.yml:52]: We are dropping the bg jobs images, so only build the `web` image here.

File: .github/workflows/build.yml [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+        TERRAFORM_VAR_FILE: terraform_qa.tfvars
+        TF_VAR_ttapi_image_tag: ${{env.SHA_TAG}}
+        TF_VAR_api_url: https://api.london.cloud.service.gov.uk
+        TF_VAR_docker_image: ${{ env.DOCKER_IMAGE }}:${{ env.SHA_TAG }}
+        TF_VAR_user: ${{ secrets.TF_VAR_user }}
+        TF_VAR_password: ${{ secrets.TF_VAR_password }}
+        TF_VAR_SECRET_KEY_BASE: ${{ secrets.TF_VAR_SECRET_KEY_BASE }}
+        TF_VAR_SENTRY_DSN: ${{ secrets.TF_VAR_SENTRY_DSN }}
+        # Let's confirm if these variables are required
+        #TF_VAR_SETTINGS__GOOGLE__GCP_API_KEY: ${{ secrets.TF_VAR_SETTINGS__GOOGLE__GCP_API_KEY }}
```
Comment (misaka) [inline .github/workflows/build.yml:124]: I just had a further thought about this one ... this key is used by the geocoding bg process, I would bet. That would probably mean we'll need it once we want to get that background job working.

**Pull request merged**